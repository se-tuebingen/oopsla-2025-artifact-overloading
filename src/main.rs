use overloading::compile_expr;
use overloading::syntax::parse_expr;

use clap::Parser;

#[derive(Parser)]
#[command(name = "overloading")]
struct Args {
    /// Input file to process
    filename: String,

    /// Use short output format
    #[arg(short, long)]
    short: bool,
}

/// Main entry point for the native CLI.
fn main() {
    let args = Args::parse();

    let content = std::fs::read_to_string(&args.filename).expect("Failed to read file");
    match parse_expr(&content) {
        Ok(expr) => {
            println!("{}", compile_expr(&expr, args.short));
        }
        Err(err) => {
            panic!("Failed to parse expression: {}", err);
        }
    }
}

/// Snapshot tests for the compiler examples.
#[cfg(test)]
#[test]
fn test_compiler_examples() {
    use insta::{assert_json_snapshot, glob, with_settings};

    glob!("../", "examples/*.over", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let output = match parse_expr(&input) {
            Ok(expr) => overloading::compile_expr_json(&expr, true),
            Err(err) => overloading::json::JsonProcessingResult::InferenceError {
                expression: input,
                error: err,
            },
        };

        // Use the filename (without extension) as snapshot suffix
        let filename = path.file_stem().unwrap().to_str().unwrap();
        with_settings!({
            snapshot_path => "../examples/snapshots",
            snapshot_suffix => filename,
            sort_maps => true,
        }, {
            assert_json_snapshot!(output);
        });
    });
}
