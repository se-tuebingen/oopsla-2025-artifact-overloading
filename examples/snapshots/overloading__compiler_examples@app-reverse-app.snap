---
source: src/main.rs
expression: output
input_file: examples/app-reverse-app.over
---
{
  "status": "success",
  "expression": "assume succInt: (Int) => Int\nlet app = λ (f, x) => a‹f(x), x(f)›\napp(succInt, 42)",
  "dimension_spans": "[spans]",
  "inferred_type": "app_5",
  "worlds": {
    "type": "one"
  },
  "solutions": [
    [
      "a₁"
    ]
  ],
  "constraints": [
    {
      "from": "f_0",
      "choices": [
        "a₁"
      ],
      "to": "(x_1) => app_3"
    },
    {
      "from": "app_3",
      "choices": [
        "a₁"
      ],
      "to": "over_a_2"
    },
    {
      "from": "x_1",
      "choices": [
        "a₂"
      ],
      "to": "(f_0) => app_4"
    },
    {
      "from": "app_4",
      "choices": [
        "a₂"
      ],
      "to": "over_a_2"
    },
    {
      "from": "(f_0, x_1) => over_a_2",
      "choices": [],
      "to": "((Int) => Int, Int) => app_5"
    }
  ],
  "bounds": [
    {
      "variable": "app_3",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "over_a_2",
          "choices": [
            "a₁"
          ]
        }
      ]
    },
    {
      "variable": "app_4",
      "lower": [],
      "upper": [
        {
          "type_name": "over_a_2",
          "choices": [
            "a₂"
          ]
        }
      ]
    },
    {
      "variable": "app_5",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": []
    },
    {
      "variable": "f_0",
      "lower": [
        {
          "type_name": "(Int) => Int",
          "choices": []
        }
      ],
      "upper": [
        {
          "type_name": "(x_1) => app_3",
          "choices": [
            "a₁"
          ]
        }
      ]
    },
    {
      "variable": "over_a_2",
      "lower": [
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ],
      "upper": [
        {
          "type_name": "app_5",
          "choices": []
        }
      ]
    },
    {
      "variable": "x_1",
      "lower": [
        {
          "type_name": "Int",
          "choices": []
        }
      ],
      "upper": [
        {
          "type_name": "(f_0) => app_4",
          "choices": [
            "a₂"
          ]
        },
        {
          "type_name": "Int",
          "choices": [
            "a₁"
          ]
        }
      ]
    }
  ],
  "errors": [
    {
      "constraint": {
        "from": "Int",
        "choices": [
          "a₂"
        ],
        "to": "(f_0) => app_4"
      },
      "message": "Unsolvable constraint (fallthrough)"
    }
  ]
}
